operator/pkg/gateway-api/cell.go

```diff
@@ -221,58 +226,131 @@ func checkCRD(ctx context.Context, clientset k8sClient.Clientset, gvk schema.Gro
+		}
+		if err := checkCRD(ctx, clientset, optionalGVK); err == nil {
+			// note that the .Kind field contains the _resource_ name -
+			// the plural, lowercase version of the name.
+			optionalGVKList := schema.GroupVersionKind{
+				Group:   optionalGVK.Group,
+				Version: optionalGVK.Version,
+				Kind:    optionalGVK.Kind[:len(optionalGVK.Kind)-1] + "lists",
+			}
+			presentGVKs = append(presentGVKs, optionalGVK, optionalGVKList)
+		} else {
+			logger.Debug("CRD is not present, will not handle it", "optionalGVK", optionalGVK.String())
+		}
+	}
+
+	return presentGVKs, res
 }
 
-// registerReconcilers registers the Gateway API reconcilers to the controller-runtime library manager.
-func registerReconcilers(mgr ctrlRuntime.Manager, translator translation.Translator, logger *slog.Logger) error {
-	reconcilers := []interface {
+// registerReconcilers registers Gateway API reconcilers to the controller-runtime library manager.
+// optionalKinds are previously autodetected based on what CRDs are present in the cluster.
+func registerReconcilers(mgr ctrlRuntime.Manager, translator translation.Translator, logger *slog.Logger, installedCRDs []schema.GroupVersionKind) error {
+	requiredReconcilers := []interface {
 		SetupWithManager(mgr ctrlRuntime.Manager) error
 	}{
 		newGatewayClassReconciler(mgr, logger),
-		newGatewayReconciler(mgr, translator, logger),
+		newGatewayReconciler(mgr, translator, logger, installedCRDs),
 		newReferenceGrantReconciler(mgr, logger),
 		newHTTPRouteReconciler(mgr, logger),
 		newGammaHttpRouteReconciler(mgr, translator, logger),
 		newGRPCRouteReconciler(mgr, logger),
-		newTLSRouteReconciler(mgr, logger),
 	}
 
-	for _, r := range reconcilers {
+	for _, r := range requiredReconcilers {
 		if err := r.SetupWithManager(mgr); err != nil {
 			return fmt.Errorf("failed to setup reconciler: %w", err)
 		}
 	}
 
+	// TODO:
+	// rename optionalGVKs to installedCRDs or something
+	// add installedCRDs support to HTTPRoute reconciler
+
+	// To add a new optionalKind, remember you also need to add the GVK into
+	// the optionalGVKs global.
+	// Note that optionalKinds contains the lower-case, plural version of the
+	// name.
+	for _, gvk := range installedCRDs {
+		if strings.HasSuffix(gvk.Kind, "lists") {
+			// skip "List" GVKs as we'll never need to handle those individually
+			continue
+		}
+		switch gvk.Kind {
+		case "tlsroutes":
+			logger.Info("TLSRoute CRD is installed, TLSRoute support is enabled")
+			tlsReconciler := newTLSRouteReconciler(mgr, logger)
+			if err := tlsReconciler.SetupWithManager(mgr); err != nil {
+				return fmt.Errorf("failed to setup optional reconciler: %w", err)
+			}
+		case "serviceimports":
+			// we don't need a reconciler, but we do need to tell folks that the
+			// support is working.
+			logger.Info("ServiceImport CRD is installed, ServiceImport support is enabled", "installedCRDs", installedCRDs)
+		default:
+			logger.Debug("No reconciler available for GVK", "GVK", gvk)
```

Comment:
IMO this would be a programmer error -> worth to return an error (or even panic)